src/ExtractTileOperations.cpp

```diff
@@ -0,0 +1,414 @@
+}
+
+Stmt convert_to_zero(const Store *op, int tile_x, int tile_y, const string &new_name) {
+    if (const auto *ramp = op->index.as<Ramp>()) {
+        if (const auto *bcast = op->value.as<Broadcast>()) {
+            if (is_const_one(ramp->stride) &&
+                is_const_zero(bcast->value) &&
+                (bcast->lanes == tile_x * tile_y)) {
+                auto rows = Cast::make(Int(16), tile_x);
+                auto bytes = op->value.type().bytes();
+                auto colbytes = Cast::make(Int(16), tile_y * bytes);
+                const auto &store_type = op->value.type();
+                // will be f32 or i32
+                auto tile_zero_type = store_type.with_lanes(1024 / store_type.bytes());
+                auto val = Call::make(tile_zero_type, "tile_zero", {rows, colbytes}, Call::Intrinsic);
+                auto store = Store::make(new_name, std::move(val), Ramp::make(0, 1, 256), Parameter(), const_true(256), ModulusRemainder());
+                return store;
+            }
+        }
+    }
+    return {};
+}
+
+Stmt convert_to_tile_store(const Store *op, const string &amx_name, int tile_x, int tile_y) {
+    auto tile = is_2d_tile_index(op->index);
+    if (tile.result && tile.extent[0] == tile_x && tile.extent[1] == tile_y) {
+        auto out = Variable::make(Handle(), op->name);
+        auto tile_type = op->value.type().with_lanes(256);
+        auto tile_val = Load::make(tile_type, amx_name, Ramp::make(0, 1, 256), {}, {}, const_true(256), {});
+        auto bytes = op->value.type().bytes();
+        internal_assert(bytes == 4) << "AMX store only supported for int32 and float32 output, not for " << op->value.type() << "\n";
+        // {tile_x, tile_y, var, base, stride}
+        auto store = Call::make(Int(32), "tile_store", {tile_x, tile_y * bytes, std::move(out), tile.base * bytes, tile.stride[0] * bytes, std::move(tile_val)}, Call::Intrinsic);
+        return Evaluate::make(std::move(store));
+    }
+    return {};
+}
+
+class ExtractTileOperations : public IRMutator {
+    using IRMutator::visit;
+
+    string tile_name;
+    string amx_name;
+    vector<Stmt> pending_stores;
+    bool in_allocate = false;
+    int found_tile_x = -1;
+    int found_tile_y = -1;
+    int found_tile_r = -1;
+    AMXOpType op_type;
+
+    Stmt visit(const Allocate *op) override {
+        if (op->memory_type == MemoryType::AMXTile) {
+            user_assert(
+                (op->type.is_int() && op->type.bits() == 32) ||
+                (op->type.is_float() && op->type.bits() == 32))
+                << "scheduled tile operations must yield 32-bit integers or 32-bit floats";
+
+            if (op->type.is_int() && op->type.bits() == 32) {
+                op_type = AMXOpType::Int8;
+            } else {
+                op_type = AMXOpType::Bf16;
+            }
+
+            user_assert(!in_allocate) << "Found two possible tile allocations for AMX allocation";
```

Comment:
Would it be helpful to append amx_name or tile_name to the error message, for debugging purposes?